crates/fluvio/src/producer/mod.rs

```diff
@@ -29,355 +35,301 @@ pub struct TopicProducer {
+    // we have to "subscribe" to a sender to get it
+    pub(crate) _status_sender: tokio::sync::broadcast::Sender<BatchStatus>,
+    /// Used by the producer to return batch statuses to the caller.
+    pub(crate) status_receiver: Mutex<tokio::sync::broadcast::Receiver<BatchStatus>>,
+}
+
+impl Drop for ProducerInner {
+    fn drop(&mut self) {
+        debug!("Notify dispatcher shutdown");
+        self.dispatcher_shutdown.notify();
+    }
 }
 
 impl TopicProducer {
-    pub(crate) fn new(topic: String, pool: Arc<SpuPool>) -> Self {
+    pub(crate) fn new(topic: String, pool: Arc<SpuPool>, config: ProducerConfig) -> Self {
         let partitioner = Box::new(SiphashRoundRobinPartitioner::new());
+
+        let (to_dispatcher, from_dispatcher) = flume::bounded(0);
+        let (status_sender, status_receiver) = tokio::sync::broadcast::channel(100);
+        let dispatcher =
+            Dispatcher::new(pool.clone(), status_sender.clone(), from_dispatcher, config);
+        let dispatcher_shutdown = dispatcher.start();
+        let inner = Arc::new(ProducerInner {
+            dispatcher: to_dispatcher,
+            dispatcher_shutdown,
+            uid_generator: RecordUidGenerator::new(),
+            _status_sender: status_sender,
+            status_receiver: Mutex::new(status_receiver),
+        });
+
         Self {
             topic,
             pool,
             partitioner,
+            inner,
         }
     }
 
     /// Sends a key/value record to this producer's Topic.
     ///
     /// The partition that the record will be sent to is derived from the Key.
     ///
+    /// The `send` function does not wait for records to be committed to the SPU,
+    /// it simply buffers them. In order to make sure that your records have all
+    /// been actually sent, be sure to call [`TopicProducer::flush`] after sending
+    /// all the records you have.
+    ///
     /// # Example
     ///
     /// ```
     /// # use fluvio::{TopicProducer, FluvioError};
     /// # async fn example(producer: &TopicProducer) -> Result<(), FluvioError> {
     /// producer.send("Key", "Value").await?;
+    /// producer.flush().await?;
     /// # Ok(())
     /// # }
     /// ```
     #[instrument(
         skip(self, key, value),
         fields(topic = %self.topic),
     )]
-    pub async fn send<K, V>(&self, key: K, value: V) -> Result<(), FluvioError>
+    pub async fn send<K, V>(&self, key: K, value: V) -> Result<RecordUid>
     where
         K: Into<RecordKey>,
         V: Into<RecordData>,
     {
-        let record_key = key.into();
-        let record_value = value.into();
-        self.send_all(Some((record_key, record_value))).await?;
-        Ok(())
+        let key = key.into();
+        let value = value.into();
+        let record = Record::from((key, value));
+
+        {
+            // Check if there have been any failures
+            let mut receiver = self.inner.status_receiver.lock().expect("Mutex poisoned");
```

Comment:
should use async lock

Reply:
I can go ahead and do this if you want me to, but I was reading about this yesterday and I think the std mutex should be fine here. We don't need to hold the lock across any `.await` points and there shouldn't be high contention. The tokio guide was basically saying to reach for std::sync::Mutex since it is faster, and only fall back to an async mutex when really needed: https://tokio.rs/tokio/tutorial/shared-state#on-using-stdsyncmutex

Let me know if you want me to switch it, but that's why I had it that way to start

Reply:
`async_mutex` doesn't use std mutex and much superior